File: TypeScript/142. Linked List Cycle II.ts
// noinspection JSUnusedGlobalSymbols
const detectCycle = (head: ListNode | null): ListNode | null => {
  if (head === null || head.next === null) {
    return null
  }

  let slow: ListNode | null | undefined = head
  let fast: ListNode | null | undefined = head

  while (fast !== null && fast.next !== null) {
    slow = slow?.next
    fast = fast.next.next
    if (slow === fast) {
      fast = head
      while (slow !== fast) {
        slow = slow?.next
        fast = fast?.next
      }

      return slow
    }
  }
  return null
}

File: Python/142. Linked List Cycle II.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# noinspection PyMethodMayBeStatic, PyPep8Naming
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                fast = head
                while slow != fast:
                    slow = slow.next
                    fast = fast.next
                return slow
        return None
